### PR TITLE
feat(trending): support nft

### PR DESCRIPTION
**Pull request by UncleBill**
https://mask.atlassian.net/browse/MF-469
https://mask.atlassian.net/browse/MF-1268

### CODE REVIEW

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/CoinMenu.tsx [hunk 1]
```diff
@@ -16,43 +29,38 @@ export interface CoinMenuOption {
...
     value: string
 }
-
-export interface CoinMenuProps {
+interface TokenMenuListProps {
     options: CoinMenuOption[]
-    selectedIndex?: number
-    onChange?: (option: CoinMenuOption) => void
-    children?: React.ReactNode
+    type?: TrendingCoinType
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trending/CoinMenu.tsx:64]: Introduce `TrendingCoinType` to distinguish between `Token` and `NFT`

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/PriceChartDaysControl.tsx [hunk 2]
```diff
@@ -31,25 +31,34 @@ export enum Days {
     ONE_DAY = 1,
     ONE_WEEK = 7,
     ONE_MONTH = 30,
+    THREE_MONTHS = 90,
     ONE_YEAR = 365,
 }
 
+export const DEFAULT_RANGE_OPTIONS = [Days.ONE_DAY, Days.ONE_WEEK, Days.ONE_MONTH, Days.ONE_YEAR, Days.MAX]
+export const NFT_RANGE_OPTIONS = [Days.ONE_DAY, Days.ONE_WEEK, Days.ONE_MONTH, Days.THREE_MONTHS]
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trending/PriceChartDaysControl.tsx:39]: Our NFT data provider NFTScan only provides `[Days.ONE_DAY, Days.ONE_WEEK, Days.ONE_MONTH, Days.THREE_MONTHS]`

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/TickersTable.tsx [hunk 3]
```diff
@@ -43,77 +45,104 @@ export interface TickersTableProps {
...
+        price: t('plugin_trader_table_price'),
+        marketplace: t('plugin_trader_marketplace'),
+        floor_price: t('plugin_trader_floor_price'),
+        sales: t('plugin_trader_sales'),
+    }
+
+    const columns: Cells[] = useMemo(() => {
+        if (isNFT) return ['marketplace', 'sales', 'volume', 'floor_price']
+        return compact(['exchange', 'pair', isUniswap ? null : 'price', 'volume', 'updated'])
+    }, [t, isNFT, isUniswap])
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trending/TickersTable.tsx]: We are only concerned about `columns`.

**Review by lelenei — DISMISSED** [review]

File: packages/dashboard/src/pages/Wallets/components/FungibleTokenTable/index.tsx [hunk 4]
```diff
@@ -93,6 +94,7 @@ export const FungibleTokenTable = memo<TokenTableProps>(({ selectedChainId }) =>
                         id: token.id,
                         name: token.name || '',
                         symbol: token.symbol || '',
+                        type: TrendingCoinType.Fungible,
```
Comment (UncleBill) [inline packages/dashboard/src/pages/Wallets/components/FungibleTokenTable/index.tsx:97]: Introduce `TrendingCoinType` to distinguish between `Token` and `NFT`

Comment (guanbinrui) [reply packages/dashboard/src/pages/Wallets/components/FungibleTokenTable/index.tsx:97]: Can we use `TokenType`?

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/TickersTable.tsx [hunk 5]
```diff
@@ -43,77 +45,104 @@ export interface TickersTableProps {
...
+    const tickerRows = tickers.map((ticker, index) => {
+        const price = ticker.price ?? ticker.floor_price
+        const volume = isNFT ? ticker.volume_24h : ticker.volume
+        const formatter = isNFT ? ethFormatter : formatCurrency
+        const marketplaceOrExchange = (
+            <TableCell className={classes.cell} colSpan={1}>
                 {ticker.logo_url ? <img className={classes.logo} src={ticker.logo_url} /> : null}
                 <span>{ticker.market_name}</span>
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trending/TickersTable.tsx]: Should we wrap it by `<Typography />`?

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/TrendingViewDeck.tsx [hunk 6]
```diff
@@ -148,6 +148,7 @@ export function TrendingViewDeck(props: TrendingViewDeckProps) {
     }, [account, trending?.coin?.symbol])
     // #endregion
 
+    const isNFT = dataProvider === DataProvider.NFTSCAN
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trending/TrendingViewDeck.tsx]: What's the relationship between token type and data provider? Does it mean only NFTSCAN returns a non-fungible token?

File: packages/web3-shared/base/src/utils/currency.ts [hunk 7]
```diff
@@ -63,3 +63,7 @@ export function formatCurrency(value: BigNumber.Value, currency = 'USD'): string
         })
         .reduce((string, part) => string + part)
 }
+
+export const ethFormatter = (value: BigNumber.Value, sign = '\u039E') => {
+    return `${new BigNumber(value).toFixed(2)} ${sign}`
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/utils/currency.ts]: I think `fromatCurrency` is the prior choice for formatting currency.

File: packages/web3-shared/evm/types/index.ts [hunk 8]
```diff
@@ -292,6 +292,12 @@ export enum NonFungibleAssetProvider {
     ZORA = 'Zora',
 }
 
+export enum NonFungibleMarketplace {
```
Comment (guanbinrui) [inline packages/web3-shared/evm/types/index.ts]: Can we replace it with `SourceType` here?

File: packages/web3-providers/src/opensea/index.ts [hunk 9]
```diff
@@ -433,4 +435,28 @@ export class OpenSeaAPI implements NonFungibleTokenAPI.Provider<ChainId, SchemaT
...
+            chainId,
+        )
+        const slug = assetContract?.collection.slug
+        if (!slug) return
+        const response = await fetchFromOpenSea<{ stats: OpenSeaCollectionStats }>(
+            urlcat('/api/v1/collection/:slug/stats', {
+                slug,
+            }),
+            chainId,
+            this._apiKey,
```
Comment (guanbinrui) [inline packages/web3-providers/src/opensea/index.ts]: We should not depend on a front-end API key anymore.

File: packages/shared/src/UI/components/NFTList/index.tsx [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+                        : undefined
+                    return (
+                        <ListItem
+                            key={token.tokenId + token.id}
+                            className={classnames(classes.nftItem, {
+                                [classes.disabled]: disabled,
+                                [classes.selected]: selected,
+                                [classes.inactive]: inactive,
+                            })}>
+                            <Link
```
Comment (guanbinrui) [inline packages/shared/src/UI/components/NFTList/index.tsx]: There is a component called `<Linking />`. It supports rendering plain text if `link` is not available.

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/TickersTable.tsx (hunk 5 above)
Comment (UncleBill) [reply packages/mask/src/plugins/Trader/SNSAdaptor/trending/TickersTable.tsx]: Done.

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/TrendingViewDeck.tsx (hunk 6 above)
Comment (UncleBill) [reply packages/mask/src/plugins/Trader/SNSAdaptor/trending/TrendingViewDeck.tsx]: Now use `coin.type` instead.

File: packages/web3-shared/base/src/utils/currency.ts (hunk 7 above)
Comment (UncleBill) [reply packages/web3-shared/base/src/utils/currency.ts]: It's different from `fromatCurrency` that the symbol is after digits.

File: packages/web3-providers/src/opensea/index.ts (hunk 9 above)
Comment (UncleBill) [reply packages/web3-providers/src/opensea/index.ts]: Would be fixed by @Lanttcat

File: packages/shared/src/UI/components/NFTList/index.tsx (hunk 10 above)
Comment (UncleBill) [reply packages/shared/src/UI/components/NFTList/index.tsx]: Done.

File: packages/web3-shared/evm/types/index.ts (hunk 8 above)
Comment (UncleBill) [reply packages/web3-shared/evm/types/index.ts]: Fixed

File: packages/mask/src/manifest.json [hunk 11]
```diff
@@ -10,6 +10,6 @@
     "browser_action": { "default_popup": "popups.html" },
     "homepage_url": "https://mask.io",
     "description": "The portal to the new & open Internet. Send encrypted message and decentralized Apps right on top of social networks.",
-    "web_accessible_resources": ["js/*", "*.png", "*.css"],
+    "web_accessible_resources": ["js/*", "*.svg", "*.png", "*.css"],
```
Comment (guanbinrui) [inline packages/mask/src/manifest.json:13]: CC @Jack-Works Does it break auto-review?

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/CoinMenu.tsx [hunk 12]
```diff
@@ -1,76 +1,170 @@
...
+    )
+
+    const menuItems = useMemo(() => {
+        const groups: Array<[type: TrendingCoinType, options: CoinMenuOption[]]> = toPairs(
+            groupBy(options, (x) => x.coin.type),
+        ).map(([type, options]) => [Number.parseInt(type, 10), options])
+
+        if (groups.length > 1) {
+            return groups.map(([type, groupOptions]) => (
+                <>
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trending/CoinMenu.tsx]: Did you lose the `key`?

File: packages/shared/src/UI/components/NFTList/index.tsx [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+                url={token.metadata?.imageURL ?? token.metadata?.imageURL}
+                tokenId={token.tokenId}
+                classes={{
+                    loadingFailImage: classes.loadingFailImage,
+                    iframe: classes.assetPlayerIframe,
+                    imgWrapper: classes.imgWrapper,
+                    wrapper: classes.wrapper,
+                }}
+            />
+            <Typography className={classes.caption}>#{token.tokenId}</Typography>
```
Comment (guanbinrui) [inline packages/shared/src/UI/components/NFTList/index.tsx]: Should format it.

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/TickersTable.tsx (hunk 3 above)
Comment (UncleBill) [reply packages/mask/src/plugins/Trader/SNSAdaptor/trending/TickersTable.tsx]:
```suggestion
    const columns: Cells[] = useMemo(() => {
        if (isNFT) return ['marketplace', 'sales', 'volume', 'floor_price']
        return compact(['exchange', 'pair', isUniswap ? null : 'price', 'volume', 'updated'])
    }, [isNFT, isUniswap])
```

File: packages/web3-shared/base/src/utils/currency.ts [hunk 14]
```diff
@@ -61,5 +78,9 @@ export function formatCurrency(value: BigNumber.Value, currency = 'USD'): string
                     return value
             }
         })
-        .reduce((string, part) => string + part)
+        .join('')
+}
+
+export const coinFormatter = (value: BigNumber.Value, sign = '\u039E') => {
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/utils/currency.ts]: Is this formatter not used anymore?

File: packages/mask/src/plugins/Trader/SNSAdaptor/trending/CoinMenu.tsx (hunk 12 above)
Comment (UncleBill) [reply packages/mask/src/plugins/Trader/SNSAdaptor/trending/CoinMenu.tsx]: Added.

File: packages/web3-shared/base/src/utils/currency.ts (hunk 14 above)
Comment (UncleBill) [reply packages/web3-shared/base/src/utils/currency.ts]: Removed.

File: packages/shared/src/UI/components/NFTList/index.tsx (hunk 13 above)
Comment (UncleBill) [reply packages/shared/src/UI/components/NFTList/index.tsx]: Done.

**Review by guanbinrui — APPROVED** [review]

File: packages/mask/src/manifest.json (hunk 11 above)
Comment (Jack-Works) [reply packages/mask/src/manifest.json:13]: Nope. Don't worry

**Pull request merged**